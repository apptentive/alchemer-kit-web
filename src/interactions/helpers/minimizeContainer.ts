export const minimizeContainer = (container: HTMLElement, position = 'corner') => {
  container.classList.toggle(`${position}--minimized`);
};

export default minimizeContainer;
