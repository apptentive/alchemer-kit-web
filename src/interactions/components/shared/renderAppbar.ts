import minimizeContainer from '../../helpers/minimizeContainer';

const renderButton = (options: { action: string; surveyTitle: string; onPress: () => void }): HTMLElement => {
  const onKeyDown = (e: KeyboardEvent) => {
    if (e.key === 'Enter' || e.key === ' ') {
      options.onPress();
    }
  };

  const button = document.createElement('div');
  button.classList.add('apptentive-appbar__action');
  button.classList.add(`apptentive-appbar__action--${options.action}`);
  button.id = `${(options.surveyTitle || '').toLowerCase().replace(/ /g, '-')}__action--${options.action}`;
  button.style.width = '32px';
  button.style.height = '32px';
  button.addEventListener('click', options.onPress);
  button.addEventListener('keydown', onKeyDown);

  button.setAttribute('role', 'button');
  button.setAttribute('aria-label', options.action);
  button.tabIndex = 0;

  console.log(button);

  return button;
};

/**
 * Renders the survey header (icon, title, close).
 * @param {HTMLElement} minimizeTarget - The DOM node that will be used when the minimize action is invoked
 * @param {string} title - The title to render in the appbar
 * @param {string} position - The position type to be used in the minimize function
 * @param {Function} onDismiss - The method to invoke when the dismiss button is clicked
 * @returns {HTMLDivElement} - The DOM Node
 */
const renderAppbar = (minimizeTarget: HTMLElement, title: string, position: string, onDismiss: VoidFunction) => {
  // NOTE: Currently unused as an icon but needed to hold shape.
  const iconElement = document.createElement('div');
  iconElement.classList.add('apptentive-appbar__icon');

  const titleElement = document.createElement('h1');
  titleElement.classList.add('apptentive-appbar__title');
  titleElement.textContent = title;
  titleElement.tabIndex = 0;

  const minimizeAction = renderButton({
    action: 'minimize',
    surveyTitle: title,
    onPress() {
      minimizeContainer(minimizeTarget, position);
      minimizeAction.classList.toggle('apptentive-appbar__action--minimize--active');
    },
  });

  const closeAction = renderButton({
    action: 'close',
    surveyTitle: title,
    onPress: onDismiss,
  });

  const actionContainer = document.createElement('div');
  actionContainer.classList.add('apptentive-appbar__actions');
  actionContainer.append(minimizeAction);
  actionContainer.append(closeAction);

  const appbar = document.createElement('div');
  appbar.classList.add('apptentive-appbar');
  appbar.append(iconElement);
  appbar.append(titleElement);
  appbar.append(actionContainer);

  return appbar;
};

export default renderAppbar;
