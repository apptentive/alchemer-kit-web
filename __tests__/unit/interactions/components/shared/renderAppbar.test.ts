import renderAppbar from '../../../../../src/interactions/components/shared/renderAppbar';
import minimizeContainer from '../../../../../src/interactions/helpers/minimizeContainer';

jest.mock('../../../../../src/interactions/helpers/minimizeContainer.ts');

const _title = 'Title';
const _position = 'corner';
const _onDismiss = jest.fn();

describe('renderAppbar', () => {
  afterEach(() => {
    document.body.innerHTML = '';
  });

  test('properly renders appbar', () => {
    const output = renderAppbar(null as any, _title, _position, _onDismiss);

    expect(output).toMatchSnapshot();
  });

  test('properly attaches dismiss event', () => {
    document.body.append(renderAppbar(null as any, _title, _position, _onDismiss));

    const dismissButton = document.querySelector<HTMLButtonElement>(
      '.apptentive-appbar .apptentive-appbar__action--close'
    )!;

    dismissButton.click();

    expect(_onDismiss).toHaveBeenCalledTimes(1);

    _onDismiss.mockClear();
  });

  test('properly attaches minimize event', () => {
    const wrapperElement = document.createElement('div');
    wrapperElement.appendChild(renderAppbar(wrapperElement, _title, _position, _onDismiss));
    document.body.append(wrapperElement);

    const minimizeButton = document.querySelector<HTMLButtonElement>(
      '.apptentive-appbar .apptentive-appbar__action--minimize'
    )!;

    minimizeButton.click();

    expect(minimizeContainer).toHaveBeenCalledTimes(1);
    expect(minimizeButton.classList.contains('apptentive-appbar__action--minimize--active')).toBe(true);
  });
});
