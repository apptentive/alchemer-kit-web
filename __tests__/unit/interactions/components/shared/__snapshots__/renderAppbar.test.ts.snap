// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderAppbar properly renders appbar 1`] = `
<div
  class="apptentive-appbar"
>
  <div
    class="apptentive-appbar__icon"
  />
  <h1
    class="apptentive-appbar__title"
    tabindex="0"
  >
    Title
  </h1>
  <div
    class="apptentive-appbar__actions"
  >
    <div
      aria-label="minimize"
      class="apptentive-appbar__action apptentive-appbar__action--minimize"
      id="title__action--minimize"
      role="button"
      style="width: 32px; height: 32px;"
      tabindex="0"
    />
    <div
      aria-label="close"
      class="apptentive-appbar__action apptentive-appbar__action--close"
      id="title__action--close"
      role="button"
      style="width: 32px; height: 32px;"
      tabindex="0"
    />
  </div>
</div>
`;
